ocaml/xapi/rpm.ml

```diff
@@ -157,9 +159,41 @@ module Pkg = struct
     | None, None ->
         EQ
 
+  let compare_version_segment s1 s2 =
+    match (s1, s2) with
+    | Int i1, Int i2 ->
+        Int.compare i1 i2 |> order_of_int
+    | Str s1, Str s2 ->
+        String.compare s1 s2 |> order_of_int
+    | Tilde, Tilde ->
+        EQ
+    | Int _, Str _ ->
+        GT
+    | Str _, Int _ ->
+        LT
+    | Tilde, _ ->
+        LT
+    | _, Tilde ->
+        GT
+
+  let split_version_string =
+    let r = Re.Posix.compile_pat {|[a-zA-Z]+|[0-9]+|~|} in
+    fun s -> s |> Re.all r |> List.map (fun g -> Re.Group.get g 0)
+
+  let normalize v =
+    let version_segment_of_string = function
+      | "~" ->
+          Tilde
+      | s -> (
+        try Int (int_of_string s) with _ -> Str s
```

Comment:
Be aware that `int_of_string "0xabc"` does not throw an exception as you might expect.,

Reply:
Yes. No need to test the string is all digits when convert string to Int as the original implementation, because the split method has promised the string is all digits or all alphabetic. There is no situation like "0b11" may mislead `int_of_string`. So just use `int_of_string` exception to  distinguish Str and Int.
Due to the function has precondition that the input is from the split result, it is not a complete function to convert `string` to `version_segment`. So I just put it in the `normalize` function scope.

Reply:
An alternative would be to use `String.sscanf "%d"`. But you are correct that the problem does not arise because the split makes sure this is only digtits or only letters.